File: src/python/cloudinary_upload_screenshots.py
#!/usr/bin/env python3
"""
Скрипт для загрузки скриншотов конкретного приложения на Cloudinary.

Для запуска:
1. Установите необходимые зависимости: pip install cloudinary python-dotenv
2. Проверьте, что файл .env содержит ваши учетные данные Cloudinary
3. Запустите скрипт: python3 src/python/cloudinary_upload_screenshots.py

Скрипт запросит ID приложения и путь к папке со скриншотами в интерактивном режиме.
"""

import os
import sys
import glob
from pathlib import Path
from dotenv import load_dotenv
import cloudinary
import cloudinary.uploader
import cloudinary.api

# Запрашиваем у пользователя ID приложения и путь к скриншотам
def get_user_input():
    print("Загрузка скриншотов на Cloudinary")
    print("=================================")
    
    app_id = input("Введите ID приложения (например, time-capsule): ")
    screenshots_path = input("Введите путь к папке со скриншотами: ")
    
    # Удаляем кавычки, если пользователь их добавил
    screenshots_path = screenshots_path.strip('"\'')
    app_id = app_id.strip('"\'')
    
    # Проверяем наличие обязательных параметров
    if not app_id:
        print("Ошибка: ID приложения не может быть пустым")
        sys.exit(1)
    
    if not screenshots_path:
        print("Ошибка: Путь к папке со скриншотами не может быть пустым")
        sys.exit(1)
    
    # Проверяем существование директории со скриншотами
    if not os.path.exists(screenshots_path):
        print(f"Ошибка: Директория {screenshots_path} не найдена")
        sys.exit(1)
    
    # Расширяем тильду в пути, если она есть
    screenshots_path = os.path.expanduser(screenshots_path)
    
    return app_id, screenshots_path

# Загружаем переменные окружения из .env файла
env_path = os.path.join(os.path.dirname(__file__), '../../src/.env')
load_dotenv(env_path)

# Конфигурация Cloudinary
cloudinary.config(
    cloud_name=os.environ.get('CLOUDINARY_CLOUD_NAME'),
    api_key=os.environ.get('CLOUDINARY_API_KEY'),
    api_secret=os.environ.get('CLOUDINARY_API_SECRET')
)

# Корневая папка в Cloudinary для всех файлов сайта
CLOUDINARY_ROOT_FOLDER = 'website'

def create_folder(folder_path):
    """Создание папки в Cloudinary."""
    print(f"Создание папки {folder_path} в Cloudinary...")
    try:
        result = cloudinary.api.create_folder(folder_path)
        print(f"Папка {folder_path} создана успешно")
        return result
    except Exception as e:
        # Если папка уже существует, это не ошибка
        if "already exists" in str(e):
            print(f"Папка {folder_path} уже существует")
            return {"success": True}
        print(f"Ошибка при создании папки {folder_path}: {e}")
        return None

def upload_file(file_path, cloudinary_path):
    """Загрузка одного файла в Cloudinary."""
    # Разделяем путь на папку и имя файла
    if '/' in cloudinary_path:
        folder = f"{CLOUDINARY_ROOT_FOLDER}/{os.path.dirname(cloudinary_path)}"
        file_name = os.path.basename(cloudinary_path)
    else:
        folder = CLOUDINARY_ROOT_FOLDER
        file_name = cloudinary_path
    
    print(f"Загрузка {file_path} на Cloudinary в папку {folder} с именем {file_name}...")
    
    try:
        result = cloudinary.uploader.upload(
            file_path,
            folder=folder,
            public_id=file_name,
            overwrite=True,
            use_filename=False,
            unique_filename=False
        )
        print(f"Успешно загружено: {result['secure_url']}")
        return result
    except Exception as e:
        print(f"Ошибка при загрузке {file_path}: {e}")
        return None

def upload_app_screenshots(app_id, screenshots_path):
    """Загрузка всех скриншотов приложения."""
    try:
        # Создаем папки для приложения, если они еще не существуют
        create_folder(CLOUDINARY_ROOT_FOLDER)
        create_folder(f"{CLOUDINARY_ROOT_FOLDER}/apps")
        create_folder(f"{CLOUDINARY_ROOT_FOLDER}/apps/{app_id}")
        
        # Получаем список файлов изображений в директории
        image_extensions = ['.jpg', '.jpeg', '.png', '.gif']
        image_files = []
        
        for ext in image_extensions:
            image_files.extend(glob.glob(os.path.join(screenshots_path, f"*{ext}")))
            image_files.extend(glob.glob(os.path.join(screenshots_path, f"*{ext.upper()}")))
        
        print(f"Найдено {len(image_files)} изображений для загрузки")
        
        # Сортируем файлы по имени
        image_files.sort()
        
        # Отделяем обычные скриншоты от темных
        regular_screenshots = []
        dark_screenshots = []
        
        for file_path in image_files:
            file_name = os.path.basename(file_path).lower()
            if 'dark' in file_name:
                dark_screenshots.append(file_path)
            else:
                regular_screenshots.append(file_path)
        
        # Загружаем каждый скриншот
        for i, file_path in enumerate(regular_screenshots):
            screen_number = i + 1
            cloudinary_file_name = f"app-screen-{screen_number}"
            upload_file(file_path, f"apps/{app_id}/{cloudinary_file_name}")
        
        for i, file_path in enumerate(dark_screenshots):
            screen_number = i + 1
            cloudinary_file_name = f"app-screen-{screen_number}-dark"
            upload_file(file_path, f"apps/{app_id}/{cloudinary_file_name}")
        
        print("\nВсе скриншоты успешно загружены!")
        print(f"\nЧтобы использовать эти скриншоты, обновите значение screenshots в apps.json для приложения {app_id}:")
        print(f"\nНапример: [{', '.join(str(i+1) for i in range(len(regular_screenshots)))}]")
        
    except Exception as e:
        print(f"Ошибка при загрузке скриншотов: {e}")

# Запускаем загрузку скриншотов
if __name__ == "__main__":
    # Запрашиваем у пользователя параметры
    app_id, screenshots_path = get_user_input()
    print(f"\nЗагрузка скриншотов для приложения: {app_id}")
    print(f"Из папки: {screenshots_path}\n")
    
    # Запрашиваем подтверждение
    confirm = input("Продолжить? (y/n): ")
    if confirm.lower() != 'y':
        print("Операция отменена")
        sys.exit(0)
        
    # Запускаем загрузку
    upload_app_screenshots(app_id, screenshots_path) 

File: js/cloudinary-upload-screenshots.js
/**
 * Скрипт для загрузки скриншотов конкретного приложения на Cloudinary
 * Для запуска:
 * 1. Установите необходимые зависимости: npm install cloudinary dotenv fs-extra readline-sync
 * 2. Проверьте, что файл .env содержит ваши учетные данные Cloudinary
 * 3. Запустите скрипт: 
 *    - Интерактивно: node js/cloudinary-upload-screenshots.js
 *    - С аргументами: node js/cloudinary-upload-screenshots.js <app-id> <path-to-screenshots>
 * 
 * Пример: node js/cloudinary-upload-screenshots.js time-capsule ~/Desktop/time-capsule-screenshots
 */

const fs = require('fs-extra');
const path = require('path');
const readlineSync = require('readline-sync');
require('dotenv').config({ path: path.join(__dirname, '../src/.env') });
const cloudinary = require('cloudinary').v2;

// Функция для получения ввода пользователя или использования аргументов командной строки
function getInputOrArgs() {
    // Получаем аргументы командной строки
    const args = process.argv.slice(2);
    
    // Если предоставлены оба аргумента, используем их
    if (args.length >= 2) {
        const appId = args[0];
        let screenshotsPath = args[1];
        
        // Удаляем кавычки, если пользователь их добавил
        screenshotsPath = screenshotsPath.trim().replace(/^["']|["']$/g, '');
        
        // Расширяем тильду в пути, если она есть (для macOS/Linux)
        if (screenshotsPath.startsWith('~')) {
            screenshotsPath = path.join(process.env.HOME, screenshotsPath.slice(1));
        }
        
        console.log(`\nИспользуются аргументы командной строки`);
        console.log(`Приложение: ${appId}`);
        console.log(`Путь к скриншотам: ${screenshotsPath}\n`);
        
        return { appId, screenshotsPath };
    }
    
    // Иначе запрашиваем у пользователя
    console.log('Загрузка скриншотов на Cloudinary');
    console.log('=================================');
    
    const appId = readlineSync.question('Введите ID приложения (например, time-capsule): ');
    let screenshotsPath = readlineSync.question('Введите путь к папке со скриншотами: ');
    
    // Удаляем кавычки, если пользователь их добавил
    screenshotsPath = screenshotsPath.trim().replace(/^["']|["']$/g, '');
    
    // Проверяем наличие обязательных параметров
    if (!appId) {
        console.error('Ошибка: ID приложения не может быть пустым');
        process.exit(1);
    }
    
    if (!screenshotsPath) {
        console.error('Ошибка: Путь к папке со скриншотами не может быть пустым');
        process.exit(1);
    }
    
    // Расширяем тильду в пути, если она есть (для macOS/Linux)
    if (screenshotsPath.startsWith('~')) {
        screenshotsPath = path.join(process.env.HOME, screenshotsPath.slice(1));
    }
    
    return { appId, screenshotsPath };
}

// Конфигурация Cloudinary
cloudinary.config({
  cloud_name: process.env.CLOUDINARY_CLOUD_NAME,
  api_key: process.env.CLOUDINARY_API_KEY,
  api_secret: process.env.CLOUDINARY_API_SECRET
});

// Корневая папка в Cloudinary для всех файлов сайта
const CLOUDINARY_ROOT_FOLDER = 'website';

// Функция для создания папки в Cloudinary
async function createFolder(folderPath) {
  console.log(`Создание папки ${folderPath} в Cloudinary...`);
  try {
    const result = await cloudinary.api.create_folder(folderPath);
    console.log(`Папка ${folderPath} создана успешно`);
    return result;
  } catch (error) {
    // Если папка уже существует, это не ошибка
    if (error.error && error.error.message === 'Folder already exists') {
      console.log(`Папка ${folderPath} уже существует`);
      return { success: true };
    }
    console.error(`Ошибка при создании папки ${folderPath}:`, error);
    return null;
  }
}

// Функция для загрузки одного файла
async function uploadFile(filePath, cloudinaryPath) {
  // Разделяем путь на папку и имя файла
  const lastSlashIndex = cloudinaryPath.lastIndexOf('/');
  const folder = lastSlashIndex !== -1 
    ? `${CLOUDINARY_ROOT_FOLDER}/${cloudinaryPath.substring(0, lastSlashIndex)}` 
    : CLOUDINARY_ROOT_FOLDER;
  const fileName = lastSlashIndex !== -1 
    ? cloudinaryPath.substring(lastSlashIndex + 1) 
    : cloudinaryPath;
  
  console.log(`Загрузка ${filePath} на Cloudinary в папку ${folder} с именем ${fileName}...`);
  
  try {
    const result = await cloudinary.uploader.upload(filePath, {
      folder: folder,
      public_id: fileName,
      overwrite: true,
      use_filename: false,
      unique_filename: false
    });
    console.log(`Успешно загружено: ${result.secure_url}`);
    return result;
  } catch (error) {
    console.error(`Ошибка при загрузке ${filePath}:`, error);
    return null;
  }
}

// Функция для загрузки всех скриншотов приложения
async function uploadAppScreenshots(appId, screenshotsPath) {
  try {
    // Создаем папку для приложения, если она еще не существует
    await createFolder(CLOUDINARY_ROOT_FOLDER);
    await createFolder(`${CLOUDINARY_ROOT_FOLDER}/apps`);
    await createFolder(`${CLOUDINARY_ROOT_FOLDER}/apps/${appId}`);
    
    // Получаем список файлов в директории
    const files = await fs.readdir(screenshotsPath);
    
    // Фильтруем только изображения
    const imageExtensions = ['.jpg', '.jpeg', '.png', '.gif'];
    const imageFiles = files.filter(file => {
      const ext = path.extname(file).toLowerCase();
      return imageExtensions.includes(ext);
    });
    
    console.log(`Найдено ${imageFiles.length} изображений для загрузки`);
    
    // Сортируем файлы по имени
    imageFiles.sort();
    
    // Отделяем обычные скриншоты от темных
    const regularScreenshots = [];
    const darkScreenshots = [];
    
    for (const file of imageFiles) {
      if (file.toLowerCase().includes('dark')) {
        darkScreenshots.push(file);
      } else {
        regularScreenshots.push(file);
      }
    }
    
    // Загружаем каждый обычный скриншот
    for (let i = 0; i < regularScreenshots.length; i++) {
      const file = regularScreenshots[i];
      const filePath = path.join(screenshotsPath, file);
      const screenNumber = i + 1;
      const cloudinaryFileName = `app-screen-${screenNumber}`;
      
      await uploadFile(filePath, `apps/${appId}/${cloudinaryFileName}`);
    }
    
    // Загружаем каждый темный скриншот
    for (let i = 0; i < darkScreenshots.length; i++) {
      const file = darkScreenshots[i];
      const filePath = path.join(screenshotsPath, file);
      const screenNumber = i + 1;
      const cloudinaryFileName = `app-screen-${screenNumber}-dark`;
      
      await uploadFile(filePath, `apps/${appId}/${cloudinaryFileName}`);
    }
    
    console.log('\nВсе скриншоты успешно загружены!');
    console.log(`\nЧтобы использовать эти скриншоты, обновите значение screenshots в apps.json для приложения ${appId}:`);
    console.log(`\nНапример: [${Array.from({length: regularScreenshots.length}, (_, i) => i + 1).join(', ')}]`);
    
  } catch (error) {
    console.error('Ошибка при загрузке скриншотов:', error);
  }
}

// Запускаем основную функцию
async function main() {
  try {
    // Запрашиваем у пользователя параметры или используем аргументы командной строки
    const { appId, screenshotsPath } = getInputOrArgs();
    
    console.log(`\nЗагрузка скриншотов для приложения: ${appId}`);
    console.log(`Из папки: ${screenshotsPath}\n`);
    
    // Проверяем существование директории со скриншотами
    if (!await fs.exists(screenshotsPath)) {
      console.error(`Директория ${screenshotsPath} не найдена`);
      process.exit(1);
    }
    
    // Запрашиваем подтверждение, если скрипт запущен в интерактивном режиме
    if (process.argv.length < 4) {
      const confirm = readlineSync.question('Продолжить? (y/n): ');
      if (confirm.toLowerCase() !== 'y') {
        console.log('Операция отменена');
        process.exit(0);
      }
    }
    
    // Запускаем загрузку
    await uploadAppScreenshots(appId, screenshotsPath);
  } catch (error) {
    console.error('Произошла ошибка:', error);
    process.exit(1);
  }
}

// Запускаем скрипт
main(); 